### PR TITLE
Vectorize parameter interpolation in `setup_intermediate_state`

### PR DESCRIPTION
**Pull request by mcwitt**
Improves the performance of `setup_intermediate_state` roughly 10x by vectorizing the computation of intermediate parameters using `jax.vmap`.

This addresses a pretty annoying pain point where generating a set of ~50 intermediates can take > 1 minute, which slows down interactive exploration of parameter interpolation schemes. With this change it should only take a few seconds.

### CODE REVIEW

**Comment by proteneer** [comment]
Might be good to have the old implementation be turned into a test (to check that we're still getting the same parameters after this refactor).

**Comment by mcwitt** [comment]
> Might be good to have the old implementation be turned into a test (to check that we're still getting the same parameters after this refactor).

Good call; added in https://github.com/proteneer/timemachine/pull/1211/commits/723e616fcf577dbdadd9957c8acaa0f941456613.

**Review by maxentile — APPROVED** [review]
Nice!

**Review by proteneer — APPROVED** [review]
nice!

File: timemachine/fe/single_topology.py [hunk 1]
```diff
@@ -998,28 +999,42 @@ def _setup_intermediate_nonbonded_term(
...
+                    jnp.concatenate((src_qlj, jnp.full((n_pairs, 1), interpolate_w_coord(0, cutoff, lamb))), axis=1),
+                    jnp.concatenate(
+                        (
+                            interpolate_qlj_fn(src_qlj, dst_qlj, lamb),
+                            jax.vmap(interpolate_w_coord, (0, 0, None))(src_w, dst_w, lamb),
+                        ),
+                        axis=1,
+                    ),
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/single_topology.py:1031]: +1 to removing compound statements

File: timemachine/fe/single_topology.py [hunk 2]
```diff
@@ -998,28 +999,42 @@ def _setup_intermediate_nonbonded_term(
...
-            pair_idxs.append(idxs)
-            pair_params.append(new_params)
+        pair_idxs = np.array([x for x, _, _ in pair_idxs_and_params], dtype=np.int32)
+
+        if pair_idxs_and_params:
+            src_params = jnp.array([x for _, x, _ in pair_idxs_and_params])
+            dst_params = jnp.array([x for _, _, x in pair_idxs_and_params])
+
+            src_qlj, src_w = src_params[:, :3], src_params[:, 3:]
+            dst_qlj, dst_w = dst_params[:, :3], dst_params[:, 3:]
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/fe/single_topology.py:1010]: This change was part of a "simplification" to keep `w` as a column vector, which avoids the need to reshape later. I'm a bit on the fence about it too. Can revert if it seems confusing, or maybe `src_params[:, 3:4]` would be preferable?

[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/fe/single_topology.py]: Changed to create a new axis as needed in https://github.com/proteneer/timemachine/pull/1211/commits/91db3823c551b144b2f5cab8814669cc351b126b

File: timemachine/fe/single_topology.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/fe/single_topology.py]: Split out into named subexpressions in https://github.com/proteneer/timemachine/pull/1211/commits/ee614c7c01fbccd118033f5a8b3187d89137681e.
